### DISCUSSION TITLE
understanding batches on tutorial example

### DISCUSSION THREAD

**Discussion by jakubMitura14** [Project-MONAI/MONAI#4509] [Q&A] [answered]
Hello I have simple question that I can not understand in:
 https://github.com/Project-MONAI/tutorials/blob/main/3d_segmentation/spleen_segmentation_3d_lightning.ipynb.

Till this moment I mostly use single image in a batch and all was well, now I want to use multiple images in  batches.
First obviously I need to crop data using for example RandCropByPosNegLabeld - I get the subsections of the images of fixed dimensions- now time to modify dataloader:
one need to set batch size - let's say 2 as in tutorial and optionally collate_fn to your implementation
so when iterating over train loader i would expect a list with 2 dictionaries as entries or one dictionary with 4 dimensional image and label

And now I do not understand  first in tutorial we have batch size 2 - but in line
```
"images, labels = batch["image"], batch["label"]" 
```
We access it as a dictionary not as list of dictionaries - also we put it into 3d model in , so neither it is list nor 4d images
```
  output = self.forward(images)
```
As far as I understand the idea of the batch it should have 4 dimensions (first with size equal to number of batches next 3 as defined in RandCropByPosNegLabeld)  so Unet with 
```
spatial_dims=3
```
should not work.

Yet all works, thanks to your work :), so where is my mistake in understanding?

Thanks for time reading it!

**Comment by rijobro** [comment, marked as answer]
The default collate function will combine elements of the same key. In this way, you'll have a single dictionary, and each value in that dictionary will be a batch. So `batch["image"]` will be a `torch.Tensor`, but will now have the shape `B,C,H,W,[D]`. The added `B` is the size of the batch, so in your example will be `2`.

**Reply by jakubMitura14** [reply]
Thank You !
